server/src/main/java/org/elasticsearch/action/ingest/ReservedPipelineAction.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.ingest;
+
+import org.elasticsearch.ElasticsearchGenerationException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoRequest;
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoResponse;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.internal.node.NodeClient;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.ingest.IngestMetadata;
+import org.elasticsearch.ingest.IngestService;
+import org.elasticsearch.reservedstate.ReservedClusterStateHandler;
+import org.elasticsearch.reservedstate.TransformState;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * This Action is the reserved state save version of RestPutPipelineAction/RestDeletePipelineAction
+ * <p>
+ * It is used by the ReservedClusterStateService to add/update or remove ingest pipelines. Typical usage
+ * for this action is in the context of file based state.
+ */
+public class ReservedPipelineAction implements ReservedClusterStateHandler<ReservedPipelineAction.ReservedPipelinesData> {
+    public static final String NAME = "ingest_pipelines";
+
+    private final IngestService ingestService;
+    private final NodeClient nodeClient;
+
+    /**
+     * Creates a ReservedPipelineAction
+     *
+     * @param ingestService requires {@link IngestService} for storing/deleting the pipelines
+     */
+    public ReservedPipelineAction(IngestService ingestService, NodeClient nodeClient) {
+        this.ingestService = ingestService;
+        this.nodeClient = nodeClient;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    public Collection<PutPipelineRequest> prepare(ReservedPipelinesData pipelines) {
+        var exceptions = new ArrayList<String>();
+        for (var pipeline : pipelines.requests) {
+            try {
+                ingestService.validatePipelineRequest(pipeline, pipelines.nodesInfos);
+            } catch (Exception e) {
+                exceptions.add(e.getMessage());
```

Comment:
Here we collect validation errors and collapse them into a single exception, but I notice in the SLM operator PR we are piecemeal in that the first exception halts execution. Should we add the collection of multiple failures to the other PR?

Reply:
Ah yeah, sorry, I forgot about that. It's better user experience to collect all errors. I'll fix that for SLM.